### ISSUE TITLE
Problem with FBT and FF pref "layout.word_select.stop_at_punctuation"

### ISSUE THREAD

**Issue by phorham** [Quicksaver/FindBar-Tweak#244]
Using: Firefox 44.0 and FindBar Tweak 2.1.8 running on openSUSE Tumbleweed

If I set the Firefox preference "layout.word_select.stop_at_punctuation" to "True" (which I require)
FindBar Tweak is resetting this to "False" when Firefox restarts.

I noticed also that when I set that preference it also results in (the FindBar Tweak) preference "extensions.findbartweak.layoutStopAtPunctuation" changing to True.

The only FindBar Tweak setting I've found that appears related is:
Preferences -> Tweaks -> "Double-clicking on word stops selection at punctuation"

If I enable that setting, it too is disabled (un-checked) after a restart of Firefox.

If I temporarily disable FindBar Tweak, when I set the Firefox preference "layout.word_select.stop_at_punctuation" to "True" that setting correctly remains when Firefox restarts.

Basically what I want is the Firefox preference "layout.word_select.stop_at_punctuation" set to "True" without FindBar Tweak setting it to False when Firefox restarts.

Am I missing something blatantly obvious, or is this a FindBar Tweak bug?

**Comment by Quicksaver** [comment]
Preferences -> Tweaks -> "Double-clicking on word stops selection at punctuation" is merely a shortcut to "layout.word_select.stop_at_punctuation". "extensions.findbartweak.layoutStopAtPunctuation" is the add-on's intermediary, basically to allow it to restore the previous value of the original preference when the add-on is uninstalled (a requirement by add-on policy).

But of course that it should persist after restarting the browser, I don't understand why it wouldn't. Please try pressing the "Reset all changed native preferences" button (in the Tweaks pane in FBT's preferences) and see if that helps with your case.

If that doesn't help, can you also please try in a [clean profile](https://support.mozilla.org/kb/profile-manager-create-and-remove-firefox-profiles) with only FBT installed?

**Comment by Quicksaver** [comment]
> Preferences -> Tweaks -> "Double-clicking on word stops selection at punctuation" is merely a shortcut to "layout.word_select.stop_at_punctuation". "extensions.findbartweak.layoutStopAtPunctuation" is the add-on's intermediary, basically to allow it to restore the previous value of the original preference when the add-on is uninstalled (a requirement by add-on policy).

This was all to say, all of those are equivalent, changing one should also change all the others so that's expected.

**Comment by phorham** [comment]
First test (Reset):

After "Reset all changed native preferences" (Reset them to their default values now)

"layout.word_select.stop_at_punctuation" False
"extensions.findbartweak.layoutStopAtPunctuation" False

Change "layout.word_select.stop_at_punctuation" to "True"
"extensions.findbartweak.layoutStopAtPunctuation" also changes to "True"

Close Firefox, restart Firefox

"layout.word_select.stop_at_punctuation" False
"extensions.findbartweak.layoutStopAtPunctuation" False

This is still the same (erroneous?) behaviour as I first reported. FBT is resetting "layout.word_select.stop_at_punctuation" to False.


Second test (New FF profile)

With Firefox closed temporarily renamed ".mozilla" 

Started Firefox (which now has a completely new default profile).

Installed FindBar Tweak (from https://addons.mozilla.org/en-US/firefox/addon/findbar-tweak).

All FindBar Tweak settings are defaults.

Closed and Restarted Firefox

"layout.word_select.stop_at_punctuation" False
"extensions.findbartweak.layoutStopAtPunctuation" False

Change "layout.word_select.stop_at_punctuation" to "True"
"extensions.findbartweak.layoutStopAtPunctuation" also changes to "True"

Close Firefox, restart Firefox

"layout.word_select.stop_at_punctuation" False
"extensions.findbartweak.layoutStopAtPunctuation" False

This again is the *identical* behaviour. FBT is still resetting "layout.word_select.stop_at_punctuation" to False.

**Comment by Quicksaver** [comment]
That is so weird, it shouldn't be happening at all. It works fine for me with the latest beta, can you give [beta version 2.1.9b2](https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.1.9b2) a try? Note that it's not signed, so you'll have disable ```xpinstall.signatures.required``` to install it.

Can you also check if the same happens with these preferences:
- FBT's preferences -> General -> "Search for text when I start typing" / ```accessibility.typeaheadfind```
- FBT's preferences -> Tweaks -> "Double-clicking on a word also selects the following space" / ```layout.word_select.eat_space_to_next_word```

All of those together with ```stop_at_punctuation``` are just shortcuts to the native preferences, all that changes between them is the preference name, so I'd like to know if this is specific to ```stop_at_punctuation``` or if it's a general problem with all the prefs of that kind.

**Comment by phorham** [comment]
OK Thanks. I'll try that, but it won't now be until tomorrow.

What I can add is I've also another computer which is running openSUSE 13.2 / Firefox 38.0.6 / FBT 2.0.6 ... and that also exhibits this behaviour.

As you're not seeing this behaviour, could it be platform dependant perhaps? I don't know what platform you're running on, I'm using, and only able to test, on linux.

**Comment by Quicksaver** [comment]
> As you're not seeing this behaviour, could it be platform dependant perhaps?

That was my thinking. It could be, but the only thing in that part of the code that changes is only the default value of the preference. Forcing the default to false on my side doesn't change a thing, I can toggle the preferences as I want them and they stay changed after restarting Firefox. I will try on linux later today anyway to make sure though.

**Comment by Quicksaver** [comment]
By the way, that's also why I asked about the other preferences, ```accessibility.typeaheadfind``` defaults to false in all operating systems as well, and that is toggling fine for me as well.

**Comment by Quicksaver** [comment]
I just checked on Ubuntu and it works fine for me as well, I tried several combinations of this preference and tried toggling it through about:config and through FBT's preferences tab.

I also checked Firefox's code, that preference is enabled by default in every OS, and it never (un)sets that preference by itself. FBT never (re)sets this preference when directly shutting down / restarting Firefox, only when it is disabled/uninstalled manually. Even if this was the case, it would still set it back to whatever you defined after starting Firefox up again.

Maybe it could be a conflict with another add-on, have you tried disabling every other add-on except for FBT? This could have been also the case in the clean profile, some linux distros like to pre-install a few integration add-ons. Ubuntu has like four of these for instance, don't know about openSUSE though as I've never tried it. It could even be other running applications pre-installing their add-ons into the clean profile.

**Comment by phorham** [comment]
OK Thanks... All your comments noted... Very weird indeed.


Testing with FF 44.0 and FBT 2.1.9b2

A new FF profile was used, FBT is the only extension.

After using FF's config editor to set "layout.word_select.stop_at_punctuation" and "layout.word_select.eat_space_to_next_word" to "True" these are both being reset to "False" upon closing/restarting FF

"accessibility.typeaheadfind" remains as set (to either "True" or "False")

The information about FF (from "about:support"): http://paste.opensuse.org/view/raw/5a3514c5

Whilst testing the above I saved the FF prefs.js at various stages.

prefs.js.001 http://paste.opensuse.org/view/raw/d45afb82 - After a clean install of FF

prefs.js.002 http://paste.opensuse.org/view/raw/f8090e10 - After "xpinstall.signatures.required" set "True" and FBT v2.1.9b2 installed

prefs.js.003 http://paste.opensuse.org/view/raw/85a237e3 - After setting "... stop_at_punctuation", "... eat_space ...", and "... typeaheadfind" to "True"

The above seems very odd, at least to me :) the FBT prefs user_pref("extensions.findbartweak.layoutEatSpaces", true); and 
user_pref("extensions.findbartweak.layoutStopAtPunctuation", true); are shown, but not "layout.word_select.stop_at_punctuation" or "layout.word_select.eat_space_to_next_word", which are the two I changed with the config editor.  However using "about:config" the latter are shown as "True".

prefs.js.004 http://paste.opensuse.org/view/raw/81b90ec2 - After closing FF

prefs.js.005 http://paste.opensuse.org/view/raw/cc900fbe - After restarting FF

I appreciate that you would have difficulty in diagnosing a bug that you are unable to reproduce. So I think we may have to just pass on this one :)

**Comment by Quicksaver** [comment]
Thank you, that report really helped narrow it down! Which honestly is surprising because nothing in those files actually pointed to the cause, it just gave me an idea. But results are results I guess. :)

So turns out [openSUSE does change that default](https://build.opensuse.org/package/view_file/openSUSE:Evergreen:11.2/MozillaFirefox-branding-openSUSE/all-openSUSE.js?expand=1).

If I change that default in my profile in the same way openSUSE does, I can still toggle it to true, and it sticks after I restart Firefox. So I'm still not sure exactly why it's not sticking for you. But I think it may have to do with Firefox and the add-on and that openSUSE file conflicting with setting defaults at various times during startup. So I made a tiny change to the code to see if it helps with that.

Could you try [beta version 2.1.9b4](https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.1.9b4) and see if the preference sticks after a restart now?

**Comment by phorham** [comment]
Success... Many thanks indeed :)

Using: Firefox 44.0 / FindBar Tweak 2.1.9b4

Both and "layout.word_select.eat_space_to_next_word" now retain their settings across FF restarts. (Just for fun I tried all four possible combinations.)


Just for information:
The openSUSE firefox branding file that you linked to is an old SEL version.

The latest one is here: http://download.opensuse.org/source/tumbleweed/repo/oss/suse/src/MozillaFirefox-branding-openSUSE-40-3.1.src.rpm (Although still the same with regard to setting "layout.word_select.stop_at_punctuation" False)

Most openSUSE users would be using the "MozillaFirefox-branding-openSUSE" package, but there is no guarantee of that. An alternative would be "MozillaFirefox-branding-upstream", http://download.opensuse.org/repositories/openSUSE:/Factory/standard/x86_64/MozillaFirefox-branding-upstream-44.0-2.1.x86_64.rpm

**Comment by phorham** [comment]
Hmmm... what I meant to say was:
Both "layout.word_select.stop_at_punctuation" and "layout.word_select.eat_space_to_next_word" now retain their settings across FF restarts. (Just for fun I tried all four possible combinations.)

I'll blame it on copy and paste...

**Comment by phorham** [comment]
Definitely that openSUSE branding over-ride...
On  another computer which is running openSUSE 13.2 / Firefox 38.0.6 / FBT 2.0.6 (and also exhibiting the same "bug") I edited the "/usr/lib64/firefox/defaults/pref/all-openSUSE.js" file to remove the "layout.word_select.stop_at_punctuation" entry.
That now also retains the settings across Firefox restarts.
